<div style="width:100%; overflow:hidden; background-color:#F1F1E6; padding: 10px; border-style: outset; color:#17469e">
    <div style="width: 80%; float: left;">
    <h2 align="center">Universidad de Sonora</h2>
    <hr style="border-width: 3px; border-color:#17469e">
          <h1>Reconocimiento de patrones: Introducción</h1>
          <h4>Ramón Soto C. <a href="mailto:rsotoc@moviquest.com/">(rsotoc@moviquest.com)</a></h4>
    </div>
    <div style="float: right;">
    <img src="images/escudo_unison.png">
    </div>
</div>

## Bases matemáticas

La base de las técnicas de reconocimiento de patrones es la comparación entre elementos del universo de discurso del problema y la determinación de semejanza entre ellos. Esta comparación se realiza a partir de las características de los elementos. 


### Atributos y características
Un **atributo** es una representación simbólica o numérica de una *propiedad* de un objeto, que puede ser útil para clasificarlo en una clase. Los diferentes objetos en un problema son descritos mediante diferentes *conjuntos de atributos*: En una clasificación de animales, por ejemplo, la descripción de las aves incluye atributos como el tipo de plumaje o la longitud del pico, atributos que no tiene sentido para otras clases de animales. 

Una manera de sistematizar y homogenizar la descripción de elementos en un problema es utilizando el mismo conjunto de propiedades para todos ellos, sean útiles o no para la clasificación de un objeto en particular, organizadas en lo que se denomina **vector de características**, así, podemos entender por **característica**, en el contexto de reconocimiento de patrones, cada uno de los atributos del objeto que será utilizado en su vector descriptivo: el conjunto de características es, típicamente, un subconjunto del conjunto de atributos.

Cada 'caso' representado mediante un vector de carcterísticas en el conjunto de datos se denomina **instancia** (o *ejemplo*, particularmente en el contexto de entrenamiento supervisado). 


#### Ejemplo: Base de datos de Diabetes 

* **Conjunto de datos**: *Pima Indians Diabetes* 

* **Fuente**: Lichman, M. (2013). UCI Machine Learning Repository [https://archive.ics.uci.edu/ml/datasets/Pima+Indians+Diabetes]. Irvine, CA: University of California, School of Information and Computer Science. 

* **Descripción**: El conjunto de datos *Pima Indian Diabetes Data Set* se ha convertido en un estándar para el análisis en el desempeño y precisión de métodos de diagnóstico de la diabetes, debido a que es la más amplia sistematización de observaciones de este problema. Esta base de datos reune información de mujeres indígenas proveniente de la tribu Pima residentes en Arizona y con edades igual o mayores a 21 años. La base de datos está compuesta por 768 instancias, formadas por 8 variables o atributos de entrada, además de una columna adicional que establece si la paciente tiene diabetes (1) o no (0): 

Variable | Atributos de predicción | Etiqueta 
---------| ----- | 
1 | Número de embarazos | *emb* 
2 | Concentración plasmática de glucosa a las 2 horas de una prueba de tolerancia a la glucosa oral | *gl2h* 
3 | Presión diastólica de la sangre | *pad* 
4 | Espesor de la piel del tríceps | *ept* 
5 | Cantidad de insulina en suero en dos horas | *is2h* 
6 | Índice de Masa Corporal | *imc* 
7 | Antecedentes Familiares | *fpd* 
8 | Edad | *edad* 
9 | Estado diabético | *clase* 

* **Vector de características**:  *v* = {*emb, gl2h, pad, ept, is2h, imc, fpd, edad*} 

Podemos leer y estos datos en Python, de la siguiente manera: 


In [1]:
"""
Reconocimiento de patrones: Introducción
"""

import pandas as pd

df = pd.read_csv("Data sets/Pima Indian Data Set/pima-indians-diabetes.data")
print(df.head(3))

   6  148  72  35   0  33.6  0.627  50  1
0  1   85  66  29   0  26.6  0.351  31  0
1  8  183  64   0   0  23.3  0.672  32  1
2  1   89  66  23  94  28.1  0.167  21  0


Al desplegar los datos, vemos que al no tener encabezados, el Dataframe toma como encabezados la primera línea de valores. Agregamos entonces las etiquetas como encabezados de las columnas:

In [3]:
#Lectura de los datos, agregando nombres de las columnas
df = pd.read_csv("Data sets/Pima Indian Data Set/pima-indians-diabetes.data", 
                 names = ['emb', 'gl2h', 'pad', 'ept', 'is2h', 'imc', 'fpd', 'edad', 'class'])
print(df.head(3))

   emb  gl2h  pad  ept  is2h   imc    fpd  edad  class
0    6   148   72   35     0  33.6  0.627    50      1
1    1    85   66   29     0  26.6  0.351    31      0
2    8   183   64    0     0  23.3  0.672    32      1


Para concluir la discusión de este primer ejemplo, damos un vistazo preliminar a los datos:

In [4]:
print( df.info())
print( df.describe() )

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
emb      768 non-null int64
gl2h     768 non-null int64
pad      768 non-null int64
ept      768 non-null int64
is2h     768 non-null int64
imc      768 non-null float64
fpd      768 non-null float64
edad     768 non-null int64
class    768 non-null int64
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None
              emb        gl2h         pad         ept        is2h         imc  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   
75%      6.000000

El método *info()* ofrece una primera información muy general, sobre el conjunto de datos; en este caso, básicamente que se trata de variables numérica y que el conjunto de datos contiene 768 vectores. 
El método *describe()* es más interesante: *count* nos dice cuantos registros hay de cada variable... en este caso 768 para todas las variables. *mean* es el valor promedio; por ejemplo, el promedio de embarazos en este grupo de 768 mujeres es de 3.845052. El mínimo de embarazos es 0 y el máximo 17. La cuarta parte de las mujeres han tenido 1 embarazo o menos, el 50% ha tenido 3 embarazos o menos y el 25% ha tenido 6 o más embarazos. 
Analizados con más cuidado, vemos que el índice de masa corporal (*imc*), definido como la masa del individuo entre el cuadrado de su altura, muestra como valor mínimo 0, lo cual es imposible. Esto significa que debe haber valores faltantes que en el conjunto de datos fueron registrados como 0; es decir, tenemos un problema de *datos sucios*. 

## Medidas de semejanza

La base de la comparación entre elementos de un conjunto es la determinación de la semejanza entre ellos.

![](images/people-groups.jpg)

La semejanza es un juicio subjetivo mediante el cual las personas describen su percepción de cercanía o parecido entre dos objetos. El objetivo del reconocimiento de patrones es sistematizar la medición de esta evaluación.

La teoría de la Gestalt reconoce varios principios en base a los cuales se puede explicar la percepción. Entre estos principios podemos distinguir los de "proximidad" y "similitud". 

![](images/Gestalt_grouping.jpg)
![ ](images/blank.png)

La "proximidad" se puede medir de manera directa a través de medidas de distancia. La "similitud", por su parte, es un concepto más difícil de generalizar: la similitud entre objetos puede definirse a través de diferentes características. 

![](images/similarity.png)

Las técnicas de reconocimiento de patrones miden la similitud en términos de la proximidad entre puntos de un espacio teórico. Este espacio se construye utilizando los rasgos que describen a un objeto como coordenadas, como en el siguiente ejemplo:

[![ ](images/FeatureSpace small.png)](images/FeatureSpace.png)
![ ](images/blank.png)


## Distancia(s)

La forma más común de medir la proximidad entre dos puntos es a través de la distancia. Una función de distancia es una función matemática que cumple las siguientes características:

\begin{eqnarray}
 D1 & & d(x,y)\ge 0 \\
 D2 & & d(x,y)=0 \iff x=y\\
 D3 & & d(x,y)=d(y,x)\\
 D4 & & d(x,y) + d(y,z) \ge d(x,z)
\end{eqnarray}

Existen diversas funciones de distancia, siendo algunas de las más comunes las siguientes:

### **Distancia euclidiana** 

La forma más común para medir la distancia es la llamada distancia euclidiana. Corresponde a la distancia en línea recta desde un punto al otro

![](images/euclidean.png)
![ ](images/blank.png)

En general, si $p = (p_1, p_2, ..., p_n)$ y $q = (q_1, q_2, ..., q_n)$ son dos puntos en un espacio de *n* características, la distancia euclidiana entre p y q es 

$$d(p,q) = \sqrt{(p_1-q_1)^2, (p_2-q_2)^2 \ldots (p_n-q_n)^2}\ $$ 

Por ejemplo, en Python, utilizando el paquete [*scipy.spatial.distance*](http://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist):

In [5]:
from scipy.spatial import distance

In [6]:
a = (1,2,3)
b = (4,5,6)

print('La distancia entre el punto {} y el punto {} es:'.format(a,b), distance.euclidean(a,b))

La distancia entre el punto (1, 2, 3) y el punto (4, 5, 6) es: 5.19615242271


Para calcular todas las distancias en un conjunto de datos:

In [7]:
md = distance.pdist(df, 'euclidean')
print('\nLas distancias euclidianas ({}) para los datos de diabetes son:\n'.format(md.size), md)


Las distancias euclidianas (294528) para los datos de diabetes son:
 [  66.91095707   54.29633528  115.73444431 ...,  116.46124169  116.0975663
   52.22155835]


El resultado es un conjunto de 190 datos $n(n-1)/2$, es decir, el resultado de obtener las distancias de cada renglón contra los restantes.

### **Distancia de Manhattan** 

También conocida como distancia del taxista, distancia lineal o distancia de ciudad, consiste en sumar las distancias horizontales y verticales, como lo haría un taxi que va de un punto a otro a través de las calles una ciudad. 

![](images/manhattan.png)
$$d(p,q) = \left|p-q \right| = \sum_{i=1}^n \left|p_i-q_i \right| \ $$

In [8]:
md_manhattan = distance.pdist(df, 'cityblock')
print('\nLas distancias del uber ({}) para los datos de diabetes son:\n'.format(md_manhattan.size), md_manhattan)


Las distancias del uber (294528) para los datos de diabetes son:
 [ 107.276  108.345  211.96  ...,  178.004  165.27    99.334]


### **Distancia de Minkowski** 

Es una generalización de las distancias de Manhattan $(k=1)$ y Euclidiana $(k=2)$. 

$$D(p,q)=\left(\sum_{i=1}^n \left| p_i-q_i \right|^k \right)^\frac{1}{k} $$

In [9]:
md_minkowski = distance.pdist(df, 'minkowski', 1)
print('\nLas distancias de Minkowski con k=1 para los datos de diabetes son:\n', md_minkowski)


Las distancias de Minkowski con k=1 para los datos de diabetes son:
 [ 107.276  108.345  211.96  ...,  178.004  165.27    99.334]


## Similaridad(es)

Otra forma de medir semejanza entre dos objetos es a través de una función de **similaridad**. Aunque la similaridad guarda cierta relación de *inversas* con la distancia, su naturaleza es distinta: mientras que la idea de distancia se deriva de una relación espacial, la similaridad se deriva de un conteo de rasgos similares entre dos objetos.

Una función de similaridad es una función que cumple con las siguientes condiciones:


\begin{eqnarray*}
 S1 & & s(x,y)\ge 0 \\
 S2 & & s(x,y)\le \min\{s(x,x), s(y,y)\}\\
 S3 & & s(x,y)=s(y,x)\\
 S4 & & s(x,y) + s(y,z) \le s(x,z) + s(y, y)\\
\end{eqnarray*}


siendo $x, y$ y $z$ tres elementos cualesquiera de un conjunto. 

### Similaridad y distancia

La forma más formal de obtener una función de similaridad es a través de las ya conocidas funciones de distancia. Una manera de construir la función de similaridad es definiendo $s(x,y)$ como

$$Ec1 : s(x,y) = d(x,z) + d(z,y) - d(x,y)$$

En este caso, $z$ es un valor de referencia.

Para que $s(x,y)$ definida así sea una función de similaridad debe cumplir con las condiciones $S1$ a $S4$. 

* Para verificar $S1$ reescribimos la desigualdad del triángulo:<br><br>
$d(x,z) + d(z,y) \ge d(x,y) \Longrightarrow  d(x,z) + d(z,y) - d(x,y)\ge 0$ y se satisface $S1$.<br><br>

* Un resultado que se deriva de la desigualdad del triángulo es la llamada *desigualdad inversa* en la que en lugar de establecer límites superiores se establecen límites inferiores:
<br><br>
$$d(x,y) + d(y,z) \ge d(x,z)$$
o bien...<br><br>
$$d(x,y) \ge d(x,z) − d(y,z)$$<br>
Supongamos que $d(x,z) − d(y,z) \ge 0$, entonces se cumple que <br><br>
$$d(x,y) \ge \left| d(x,z) − d(y,z)\right|$$<br>
Por otra parte, si $d(x,z) − d(y,z) < 0$, entonces $d(x,z) < d(y,z)$<br>
![](images/triangle.png)<br>
y ahora preferimos la expresión <br>
$$d(z,x) + d(x,y) \ge d(z,y)$$<br>
(que también es válida, al igual que $d(x,z) + d(z,y) \ge d(x,y)$)<br><br>
Y entonces, <br>
$$d(x,y) \ge d(z,y) − d(z,x) > 0$$<br>
y combinando ambos casos...<br>
$$d(x,y) \ge \left| d(z,y) − d(z,x)\right| = \left| d(x,z) − d(z,y)\right|$$<br>
o bien<br>
$$\left| d(x,z) - d(z,y) \right| \le d(x,y)$$<br>
por lo tanto<br>
\begin{eqnarray*}
 s(x,y) & = & d(x,z) + d(z,y) - d(x,y) \\
 s(x,y) & \le & d(x,z) + d(z,y) - \left| d(x,z) - d(z,y) \right|
\end{eqnarray*}
pero<br><br>
$\mbox{if } d(x,z) \ge d(y,z)  \Rightarrow d(x,z) + d(y,z) - \left| d(x,z) - d(y,z) \right| = d(x,z) + d(y,z) - d(x,z) + d(y,z)  = 2d(y,z)$<br><br>
y<br><br>
$\mbox{if } d(x,z) \lt d(y,z)  \Rightarrow d(x,z) + d(y,z) - \left| d(x,z) - d(y,z) \right| = d(x,z) + d(y,z) + d(x,z) - d(y,z) = 2d(x,z)$<br><br>
o bien<br>
$$
d(x,z) + d(y,z) - \left| d(x,z) - d(y,z) \right| = 
\begin{cases} 
2d(y,z), & \mbox{if } d(x,z) \ge d(y,z) \\ 
2d(x,z), & \mbox{if } d(x,z) \lt d(y,z)
\end{cases}
$$<br><br>
entonces<br><br>
$$s(x,y) \le 2\min\{d(x,z), d(y,z)\} $$<br>
y volviendo a la ecuación Ec1<br>
\begin{eqnarray*}
s(x,x) & = & 2 d(x,z)\\
s(y,y) & = & 2 d(y,z) 
\end{eqnarray*}<br>
entonces<br>
$$s(x,y) \le \min\{s(x,x), s(y,y)\} $$<br>
y se verifica la condición $S2$. <br><br>

* $S3$ es trivial<br><br>

* $S4$ se puede derivar de $D4$.<br><br>

De manera semejante, dada la función de similaridad $s(x,y)$, puede obtenerse una medida de distancia haciendo:

$$d(x, y) = s(x, x) + s(y, y)−2s(x, y)$$


### Otras alternativa

Otra alternativa más común para definir similaridades a partir de distancias es mediante:

 $$s(x,y) = \frac{1}{d(x,y)}$$

In [10]:
md = distance.pdist(df.head(5), 'euclidean')
print('\nLas distancias euclidianas para los datos de diabetes son:\n', md)

ms = 1/md
print('\nLas similaridades euclidianas para los datos de diabetes son:\n', ms)

md2 = md/10000
print('\nLas similaridades euclidianas para los datos "pequeños" de diabetes son:\n', 1/md2)


Las distancias euclidianas para los datos de diabetes son:
 [  66.91095707   54.29633528  115.73444431  172.58913326  102.52313417
   94.8171074   178.66729407  135.64400107  180.55650488   94.7391083 ]

Las similaridades euclidianas para los datos de diabetes son:
 [ 0.01494524  0.01841745  0.00864047  0.00579411  0.0097539   0.01054662
  0.005597    0.00737224  0.00553843  0.0105553 ]

Las similaridades euclidianas para los datos "pequeños" de diabetes son:
 [ 149.45235337  184.17449259   86.40470052   57.94107549   97.53896114
  105.46619987   55.96995271   73.72239039   55.38432418  105.55303063]


La siguiente modificación resuelve el problema de posible división por cero.

$$s(x,y) = \frac{1}{d(x,y)+0.5}$$

In [11]:
ms = 1/(md+0.5)
print('\nLas similaridades euclidianas para los datos de diabetes son:\n', ms)

ms2 = 1/(md2+0.5)
print('\nLas similaridades euclidianas para los datos "pequeños" de diabetes son:\n', ms2)


Las similaridades euclidianas para los datos de diabetes son:
 [ 0.01483438  0.0182494   0.0086033   0.00577737  0.00970656  0.0104913
  0.00558138  0.00734516  0.00552314  0.01049989]

Las similaridades euclidianas para los datos "pequeños" de diabetes son:
 [ 1.97358905  1.97851478  1.95475354  1.9332678   1.95981473  1.96277899
  1.93099874  1.94717547  1.93029455  1.96280904]



### Distancias normalizadas

Cuando se trabaja con similaridades puede ser conveniente que los valores obtenidos queden restringidos a un rango, típicamente en [0, 1]. De esta manera resulta claro asociar 0 a la similaridad entre objetos totalmente disímiles y 1 entre objetos idénticos. Cuando las medidas de similaridad se derivan de una distancia, es conveniente, en tales casos, normalizar la distancia. Existen diversas maneras de normalizar las distancias, siendo las más comunes las siguientes:

$$\hat{d}(x,y) = \frac{d(x,y)}{1+d(x,y)}$$

y una función de similaridad normalizada derivada de esta distancia normalizada sería:

$$s(x,y) = 1-\hat{d}(x,y)$$

In [12]:
md = distance.pdist(df.head(5), 'euclidean')
print('\nLas distancias euclidianas para los datos de diabetes son:\n', md)
mdn =  md/(1+md)
print('\nLas distancias euclidianas normalizadas para los datos de diabetes son:\n', mdn)
print('\nY las similaridades euclidianas correspondientes son:\n', 1-mdn)

md2 = md/10000
mdn2 =  md2/(1+md2)
print('\nY para los datos pequeños:\n d:', mdn2, "\n s:",1-mdn2)


Las distancias euclidianas para los datos de diabetes son:
 [  66.91095707   54.29633528  115.73444431  172.58913326  102.52313417
   94.8171074   178.66729407  135.64400107  180.55650488   94.7391083 ]

Las distancias euclidianas normalizadas para los datos de diabetes son:
 [ 0.98527484  0.98191562  0.99143355  0.99423927  0.99034032  0.98956345
  0.99443416  0.99268171  0.99449207  0.98955495]

Y las similaridades euclidianas correspondientes son:
 [ 0.01472516  0.01808438  0.00856645  0.00576073  0.00965968  0.01043655
  0.00556584  0.00731829  0.00550793  0.01044505]

Y para los datos pequeños:
 d: [ 0.00664662  0.00540031  0.01144103  0.0169661   0.01014827  0.00939265
  0.01755311  0.01338287  0.01773543  0.009385  ] 
 s: [ 0.99335338  0.99459969  0.98855897  0.9830339   0.98985173  0.99060735
  0.98244689  0.98661713  0.98226457  0.990615  ]


## Conclusiones

* En el contexto de reconocimiento de patrones, se entiende por "atributo" la representación de una propiedad de un objeto.

* Si el atributo forma parte del vector representativo del objeto, le llamamos "característica".

* Una "instancia" es cada uno de los vectores de características empleados para describir los elementos en el conjunto de observaciones.

* La comparación entre instancias en un problema se realiza mediante medidas de semejanza, que pueden ser funciones de distancia o funciones de similaridad.

<hr style="border-width: 3px;">

###### <a name="tarea_4">Tarea 4</a>

* Investigue y explique las funciones de distancia de Hamming, Pearson, Coseno y Gower. 

**Fecha de entrega**: 7 septiembre.
